keyvault/client/keys/src/main/java/com/azure/security/keyvault/keys/KeyClientBuilder.java

```diff
@@ -3,85 +3,157 @@
 
 package com.azure.security.keyvault.keys;
 
+import com.azure.core.implementation.util.ImplUtils;
+import com.azure.core.util.configuration.Configuration;
 import com.azure.core.credentials.TokenCredential;
 import com.azure.core.http.HttpClient;
 import com.azure.core.http.HttpPipeline;
+import com.azure.core.http.policy.BearerTokenAuthenticationPolicy;
 import com.azure.core.http.policy.HttpLogDetailLevel;
 import com.azure.core.http.policy.HttpLoggingPolicy;
 import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.core.http.policy.RetryPolicy;
+import com.azure.core.http.policy.UserAgentPolicy;
+import com.azure.core.util.configuration.ConfigurationManager;
 
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
 
 /**
- * This class provides a fluent builder API to help aid the configuration and instantiation of the {@link KeyClient key client},
- * calling {@link KeyClientBuilder#build() build} constructs an instance of the client.
+ * This class provides a fluent builder API to help aid the configuration and instantiation of the {@link KeyAsyncClient secret async client},
+ * calling {@link KeyClientBuilder#buildAsyncClient() buildAsyncClient} constructs an instance of the client.
+ * It also aids instantiation of the {@link KeyClient secret sync client}, calling {@link KeyClientBuilder#buildClient() buildClient}
+ * constructs an instance of the client.
  *
- * <p> The minimal configuration options required by {@link KeyClientBuilder keyClientBuilder} to build {@link KeyClient}
+ * <p> The minimal configuration options required by {@link KeyClientBuilder secretClientBuilder} to build {@link KeyAsyncClient}
  * are {@link String endpoint} and {@link TokenCredential credential}. </p>
- * <pre>
- * KeyClient.builder()
- *   .endpoint("https://myvault.vault.azure.net/")
- *   .credential(new DefaultAzureCredential())
- *   .build();
- * </pre>
+ *
+ * {@codesnippet com.azure.keyvault.keys.async.keyclient.instantiation}
  *
  * <p>The {@link HttpLogDetailLevel log detail level}, multiple custom {@link HttpLoggingPolicy policies} and custom
  * {@link HttpClient http client} can be optionally configured in the {@link KeyClientBuilder}.</p>
  * <pre>
- * KeyClient.builder()
+ * KeyAsyncClient secretAsyncClient = KeyAsyncClient.builder()
  *   .endpoint("https://myvault.vault.azure.net/")
  *   .credential(new DefaultAzureCredential())
  *   .httpLogDetailLevel(HttpLogDetailLevel.BODY_AND_HEADERS)
  *   .addPolicy(customPolicyOne)
  *   .addPolicy(customPolicyTwo)
  *   .httpClient(client)
- *   .build();
+ *   .buildAsyncClient();
  * </pre>
```

Comment:
This code snippet should be externalised - whenever you encounter one take the opportunity to extract it.